res/controllers/Behringer-CMDStudio4a-scripts.js

```diff
@@ -0,0 +1,257 @@
+    } else {
+        // No DEL so just cue.
+        engine.setValue(group, "cue_default", 1);
+    }
+}
+BehringerCMDStudio4a.cueRelease = function (channel, control, value, status, group) {
+    if (BehringerCMDStudio4a.delState[channel]) {
+        // DEL is active, reverse-roll release!
+        engine.setValue(group, "reverse", 0);
+    } else {
+        engine.setValue(group, "cue_default", 0);
+	}
+}
+
+
+// Function to deal with the scratch mode buttons.
+BehringerCMDStudio4a.scratch = function (channel, control, value, status, group) {
+    if (BehringerCMDStudio4a.scratchState[channel]) {
+        BehringerCMDStudio4a.scratchState[channel] = false;
+        midi.sendShortMsg(status, control, 0x00);
+    } else {
+        BehringerCMDStudio4a.scratchState[channel] = true;
+        midi.sendShortMsg(status, control, 0x01);
+    }
+}
+
+// Functions to deal with the hot-que buttons.
+BehringerCMDStudio4a.hotcuePush = function (channel, control, value, status, group) {
+    // Translate the button to the actual hotcue.
+    var hotcue = control-0x21;  // Hotcue buttons on left ddeck go from 0x22 to 0x29
+    if (hotcue>8) {
+        // Right deck, buttons are 0x20 higher so we need to compensate.
+        hotcue = hotcue-0x20;
+    }
+    print(hotcue)
+    if (BehringerCMDStudio4a.delState[channel]) {
+        // Del "shift" is active, delete hotcue instead.
+        engine.setValue(group, "hotcue_"+hotcue+"_clear", 1);
+    } else {
+        // Del "shift" is not active, do the push hotcue function as normal.
+        engine.setValue(group, "hotcue_"+hotcue+"_activate", 1);
+    }
+}
+BehringerCMDStudio4a.hotcueRelease = function (channel, control, value, status, group) {
+    // Translate the button to the actual hotcue.
+    var hotcue = control-0x21;  // Hotcue buttons on left ddeck go from 0x22 to 0x29
+    if (hotcue>8) {
+        // Right deck, buttons are 0x20 higher so we need to compensate.
+        hotcue = hotcue-0x20;
+    }
+    if (BehringerCMDStudio4a.delState[channel]) {
+        // Del "shift" is active, do nothing on button release.
+    } else {
+        // Del "shift" is not active, do the release hotcue function as normal.
+        engine.setValue(group, "hotcue_"+hotcue+"_activate", 0);
+    }
+}
+
+// Functions to deal with the pitch inc/dec buttons.
+BehringerCMDStudio4a.pitchDecPush = function (channel, control, value, status, group) {
+    BehringerCMDStudio4a.pitchDec[channel] = true;
+    if (BehringerCMDStudio4a.delState[channel]) {
+        // Del "shift" is active, do "key" changes instead.
+        // Check if the other button is pressed too, if so we reset the pitch.
+        if (BehringerCMDStudio4a.pitchInc[channel]) {
+            engine.setValue(group, "pitch", 0);
+        } else {
+            engine.setValue(group, "pitch_down", 1);
+        }
+    } else {
+        // DEL mode is not active, just do "rate" changes.
+        engine.setValue(group, "rate_perm_down", 1);

```

Comment:
I'd respond:

1) 10% is fine most of the time, but when you need that little bit extra it's a real pain not to have it at your fingertips, (yep. the range can be adjusted in the prefs, but then you loose precision for the majority of the time you don't need the extra range, and there is *always* going to be a time when you need "a little bit more" no matter what you set the range to).

2) Yep, zero-ing out a rate fader is a pain at the moment, but the + and - together option that I've already implemented works well enough right now. As a feature request, a customisable dead-zone (as you have with most joystick settings) really needs to go in the CC++ code I think, and then be implemented as an XML option, so I won't be trying it in JS :-).

3) Yep, I know what the "_temp" buttons are for, but your suggestion is *not* how it's done in the skins, (at least not the "LateNight" skin I'm using as a guide).
The "LateNight" skin actually has this "_perm" step functionality set-up as the +/- buttons directly under the rate faders. The "_temp" functions are actually mapped to left and right arrows lower down (under these + and - buttons), and as you can see, the CMD STUDIO 4a has +/- buttons directly under the rate faders (not arrows) so my mapping seems much more consistent with the skin-UI than your suggestion.
In any case, my point in item (3) is that you get all the "_temp" functionality you need with "_perm" mapping, plus extra benefits. Instead of holding the button down for a period to change the rate, then releasing the button to return the rate to normal (resulting in a "phase" change), with a "_perm" mapping you now tap the button to change the rate, then tap the opposite button at the end of the period to return the rate to normal. i.e. you still get the phase change, we are only exchanging a button *release* for a button *tap*. In use, you don't lose any "_temp" functionality, *and* you gain all the other benefits with this system (greater range when you need it and repeatable precision step control of rate including zero).

I really think this is a *much* better use of the deck +/- rate buttons. However, if the "_perm" button mapping really is a blocker to acceptance, let me know and I'll change it.